### PR TITLE
Feature/verify purchase

### PR DESCRIPTION
**Pull request by bizz84**

### CODE REVIEW

File: SwiftyStoreKit/InAppReceipt.swift [hunk 1]
```diff
@@ -235,5 +241,69 @@ internal class InAppReceipt {
...
+          }
+      
+          // Return the expires dates sorted desc
+          let expiresDates = receiptsMatchingProductId
+          .map { (receipt) -> NSDate in
+            // TODO: should use 'expires_date' or 'expiration_date' instead? See:
+            // https://developer.apple.com/library/mac/releasenotes/General/ValidateAppStoreReceipt/Chapters/ReceiptFields.html
+              let expires_date = receipt["expires_date_ms"] as? NSString
+              let expires_date_double = (expires_date?.doubleValue ?? 0.0) / 1000
+              return NSDate(timeIntervalSince1970: expires_date_double)
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: I'm not sure this code works as there isn't an `expires_date_ms` field [here](https://developer.apple.com/library/mac/releasenotes/General/ValidateAppStoreReceipt/Chapters/ReceiptFields.html).
In my tests I have added a consumable IAP and I have observed that after purchasing it successfully and calling the verifyPurchase() method, this call will return `NSDate(timeIntervalSince1970: 0)` which is _1 January 1970_ and means that the receipt is always expired.
By reading the docs it seems that if the `expires_date` or `expiration_date` fields are missing, the IAP can't expire and we should return `.Purchased(expiresDate: nil)`.
Instead, if we have an expiration date, we should compare it with the current date, and return `Purchased` or `Expired` based on that.
What do you think?

File: SwiftyStoreKit/SwiftyStoreKit.swift [hunk 2]
```diff
@@ -158,7 +158,26 @@ public class SwiftyStoreKit {
         password: String? = nil,
         session: NSURLSession = NSURLSession.sharedSession(),
         completion:(result: VerifyReceiptResult) -> ()) {
-            InAppReceipt.verify(receiptVerifyURL: url, password: password, session: session, completion: completion)
+        InAppReceipt.verify(receiptVerifyURL: url, password: password, session: session) { result in
+         
+            dispatch_async(dispatch_get_main_queue()) {
+                completion(result: result)
```
Comment (bizz84) [inline SwiftyStoreKit/SwiftyStoreKit.swift]: This change is technically not part of this PR but I realised it's a fix that is needed otherwise we the completion block is not called on the main thread.

File: SwiftyStoreOSXDemo/ViewController.swift [hunk 3]
```diff
@@ -32,28 +32,29 @@ class ViewController: NSViewController {
     
     // MARK: actions
     @IBAction func getInfo1(sender: AnyObject?) {
-        getInfo("1")
+        getInfo("purchase1")
     }
-    @IBAction func getInfo2(sender: AnyObject!) {
-        getInfo("2")
+    @IBAction func getInfo2(sender: AnyObject?) {
+        getInfo("consumablePurchase")
```
Comment (bizz84) [inline SwiftyStoreOSXDemo/ViewController.swift]: Registered a new `consumablePurchase` ID so that we can test consumable IAPs.

File: SwiftyStoreKit/SwiftyStoreKit.swift [hunk 4]
```diff
@@ -158,7 +158,26 @@ public class SwiftyStoreKit {
...
+    /**
+     *  Verify the purchase of a product in a receipt
+     *  - Parameter productId: the product id of the purchase to verify
+     *  - Parameter inReceipt: the receipt to test in
+     *  - Parameter validUntil: the expires date of the subscription must be valid until this date. If nil, no verification
+     */
+    public class func verifyPurchase(
+        productId productId: String,
+        inReceipt receipt: ReceiptInfo,
+        validUntil: NSDate? = nil
```
Comment (bizz84) [inline SwiftyStoreKit/SwiftyStoreKit.swift]: Shouldn't validUntil be non optional and default to `NSDate()`?

File: SwiftyStoreKit/InAppReceipt.swift (hunk 1 above)
Comment (NicolasMahe) [inline SwiftyStoreKit/InAppReceipt.swift]: The expires_date related fiels are only present for Automatically Renewable Subscription.
That's why the parameter `validUntil`is optional:
- If you want to test a consumable, non-consumable or non-renewable subscription, you should let the `validUntil` parameter to `nil`. In those case, the function return `.Purchased(expiresDate: nil)`.
- If you indicate a `validUntil`parameter to test a Automatically Renewable Subscription, the function will continue after the line 278 and test the `expires_date` with the `validUntil`.

I think the user's API should be split in 2 to avoid confusion:
- One only for Automatically Renewable Subscription with a required validUntil date.
- One for the others type of product with no validUntil parameter.
Those 2 API will use the same function in the background.


That's true that `expires_date_ms` is not in the docs, but it is present in the receipt json, as well as 'purchase_date_ms', 'original_purchase_date_ms', 'expires_date_pst', `original_purchase_date_pst` and `purchase_date_pst'. I think they didn't put them in the doc because they are just helper to parse the plain string date easier.
We can parse `expires_date` with a `NSDateFormatter` if you think it's better to use only fields present in the docs.

File: SwiftyStoreKit/SwiftyStoreKit.swift (hunk 4 above)
Comment (NicolasMahe) [inline SwiftyStoreKit/SwiftyStoreKit.swift]: See my previous response.

File: SwiftyStoreKit/InAppReceipt.swift (hunk 1 above)
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: I have updated the API so that it takes a `PurchaseType` enum as a parameter and this is used to determine whether the expiration date should be considered. This puts the responsibility on the client to decide which is the correct PurchaseType for each IAP, which I believe is a reasonable request. Should the client specify a `AutomaticallyRenewableSubscription` with a date for a purchase that really is `Consumable`, the completion block will likely say that the purchase is expired, however this can be classified as incorrect usage of the API.

File: SwiftyStoreKit/InAppReceipt.swift [hunk 5]
```diff
@@ -30,10 +30,28 @@ public typealias ReceiptInfo = [String: AnyObject]
...
+    public enum PurchaseType {
+        // A consumable In-App Purchase must be purchased every time the user downloads it.
+        case Consumable
+        // Non-consumable In-App Purchases only need to be purchased once by users.
+        case NonConsumable
+        // Automatically renewable subscriptions allow the user to purchase updating and dynamic content for a set duration of time.
+        case AutomaticallyRenewableSubscription(validUntilDate: NSDate)
+        // Free subscriptions don’t expire and can only be offered in apps that are in the Magazines & Newspapers category.
+        case FreeSubscription
+        // Non-Renewing Subscriptions allow the sale of services with a limited duration.
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: Not sure if this should also have a `validUntilDate` parameter?

File: SwiftyStoreKit/InAppReceipt.swift [hunk 6]
```diff
@@ -30,10 +30,28 @@ public typealias ReceiptInfo = [String: AnyObject]
...
     public enum VerifyReceiptResult {
         case Success(receipt: ReceiptInfo)
         case Error(error: ReceiptError)
     }
+  
+    public enum VerifyPurchaseResult {
+        case Purchased(expiresDate: NSDate?)
+        case Expired(expiresDate: NSDate)
+        case NotPurchased
+    }
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: Should we introduce separate result enums depending on whether we are dealing with Consumable/NonConsumable or Subscriptions?

File: SwiftyStoreKit/InAppReceipt.swift [hunk 7]
```diff
@@ -235,5 +253,73 @@ internal class InAppReceipt {
...
+        }
+    }
     
+    private class func verifyPurchase(receiptsMatchingProductId: [ReceiptInfo], validUntilDate date: NSDate) -> SwiftyStoreKit.VerifyPurchaseResult {
+        // Return the expires dates sorted desc
+        let expiresDates = receiptsMatchingProductId
+            .map { (receipt) -> NSDate in
+                let expires_date = receipt["expires_date_ms"] as? NSString
+                let expires_date_double = (expires_date?.doubleValue ?? 0.0) / 1000
+                return NSDate(timeIntervalSince1970: expires_date_double)
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: Is there a way that we can more gracefully handle the case where the `expires_date` value is missing and not return "1 January 1970"?

**Comment by bizz84** [comment]
@NicolasMahe See updated implementation which introduces a `PurchaseType` type to deal with different types of IAPs.
One thing I noticed is that when I run the `verifyReceipt()` method I get a `ReceiptStatus.SecretNotMatching` error which maps tho this:
```
21004
The shared secret you provided does not match the shared secret on file for your account.
Only returned for iOS 6 style transaction receipts for auto-renewable subscriptions.
```
Another possible problem is that when I restore the purchases I get multiple transactions for my auto renewable subscriptions - not sure if this is expected or not.
Have you done some testing around Subscriptions? I would like to include the `verifyPurchase()` feature with support for subscriptions however I need to understand a bit better how this works if I'm going to maintain this code going forward.

**Comment by NicolasMahe** [comment]
@bizz84 You need to indicate your Shared Secret from the iTunes Connect in-App Purchases page when you call the function `SwiftyStoreKit.verifyReceipt` in order to get the auto-renewable subscription receipt.

---

I have the same behaviour concerning the restore of auto-renewable subscriptions. When you subscribe to an auto-renewable subscription in the test endpoint environment, it is automatically renew maximum 6 times. The duration of the subscription is also different. For example, if you test a 1 month subscription, the test duration is only 5 mins. So if you restore the purchases 1h after the purchase of the first subscription, you will automatically have 5 new purchases. But the last one is already expired.
Apple established a system like these because you can't manage the subscriptions in the test environment. So these way, you can test the renew of a subscription but also its expiration.
[Apple docs](https://developer.apple.com/library/ios/documentation/LanguagesUtilities/Conceptual/iTunesConnectInAppPurchase_Guide/Chapters/TestingInAppPurchases.html)

File: SwiftyStoreKit/InAppReceipt.swift (hunk 7 above)
Comment (NicolasMahe) [inline SwiftyStoreKit/InAppReceipt.swift]: Yes, we can set the returned NSDate to an optional and check that expires_date is not missing:
```swift
let expiresDates = receiptsMatchingProductId
    .map { (receipt) -> NSDate? in
        guard let expires_date = receipt["expires_date_ms"] as? NSString
            else { return nil }
        let expires_date_double = expires_date.doubleValue / 1000
        return NSDate(timeIntervalSince1970: expires_date_double)
    }
```
We should also spread the consequences of this change to the end of the function.

File: SwiftyStoreKit/InAppReceipt.swift (hunk 6 above)
Comment (NicolasMahe) [inline SwiftyStoreKit/InAppReceipt.swift]: I think it's a good idea. The API will be even more understandable.

File: SwiftyStoreKit/InAppReceipt.swift (hunk 5 above)
Comment (NicolasMahe) [inline SwiftyStoreKit/InAppReceipt.swift]: Yes it should. The AutomaticallyRenewableSubscription need to be test with a date. If not, you will considered a expired subscription as valid.
By default the date should be `now`.

**Comment by NicolasMahe** [comment]
@bizz84 is there a way for me to push commit to these branch to help you improve the functionality ?

**Comment by bizz84** [comment]
@NicolasMahe you could try to fork this PR so that we can merge your changes here first, and later to master when we're happy that it all works.

File: SwiftyStoreKit/InAppReceipt.swift [hunk 8]
```diff
@@ -235,5 +248,101 @@ internal class InAppReceipt {
...
+            .map { (dateString) -> NSDate in
+                // If duration is set, create an "expires date" value calculated from the original purchase date
+                let addDuration = duration ?? 0
+                let expiresDateDouble = (dateString!.doubleValue / 1000 + addDuration)
+                return NSDate(timeIntervalSince1970: expiresDateDouble)
+            }
+            .sort { (a, b) -> Bool in
+                // Sort by descending date order
+                return a.compare(b) == .OrderedDescending
+            }
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: I have simplified this logic so that less map/filter methods are needed

File: SwiftyStoreKit/InAppReceipt.swift [hunk 9]
```diff
@@ -235,5 +248,101 @@ internal class InAppReceipt {
...
+      
+        guard let firstExpiryDate = expiryDateValues.first else {
+            return .NotPurchased
+        }
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: As we sort in the step above, the `firstExpiryDate` is the farthest in the future.

File: SwiftyStoreKit/InAppReceipt.swift [hunk 10]
```diff
@@ -235,5 +248,101 @@ internal class InAppReceipt {
...
+      
+        // Check if at least 1 receipt is valid
+        if firstExpiryDate.compare(date) == .OrderedDescending {
+            
+            // The subscription is valid
+            return .Purchased(expiresDate: firstExpiryDate)
+        }
+        else {
+            // The subscription is expired
+            return .Expired(expiresDate: firstExpiryDate)
```
Comment (bizz84) [inline SwiftyStoreKit/InAppReceipt.swift]: If `firstExpiryDate` is expired, any other `expiryDates` will be as well due to `.OrderDescending`

**Comment by bizz84** [comment]
@NicolasMahe I have merged your PR and simplified some of the expiry dates logic. Please review and let me know if this makes sense.
Also updated README and aligned OSX and iOS demo code.

**Comment by NicolasMahe** [comment]
Everything is ok for me 👍 
Nice modifications!
